File: src/11.binarySearch/23.483.ts
export function smallestGoodBase(n: string): string {
    const num = BigInt(n);

    const check = (x: bigint, m: bigint): number => {
        let ans = BigInt(0);
        for (let i = 0; i <= m; i++) {
            if (ans > (num - BigInt(1)) / x) {
                return 1;
            }
            ans = ans * x + BigInt(1);
        }

        if (ans === num) {
            return 0;
        } else if (ans < num) {
            return -1;
        } else {
            return 1;
        }
    };

    for (let i = BigInt(64); i > 0; i--) {
        let l = BigInt(2);
        let r = num;

        while (l < r) {
            const mid = l + (r - l) / BigInt(2);
            const temp = check(mid, i);

            if (temp === 0) {
                return mid.toString();
            } else if (temp < 0) {
                l = mid + BigInt(1);
            } else {
                r = mid;
            }
        }
    }

    return num - BigInt(1) + "";
}

File: src/11.binarySearch/23.483.go
package array_greedy

import "strconv"

func SmallestGoodBase(n string) string {
	num, _ := strconv.ParseInt(n, 10, 64)

	check := func(x, m int64) int {
		ans := int64(0)
		for i := int64(0); i < m; i++ {
			if ans > (num-1)/x {
				return 1
			}
			ans = ans*x + 1
		}
		if ans == num {
			return 0
		} else if ans < num {
			return -1
		} else {
			return 1
		}
	}

	for i := int64(64); i > 0; i-- {
		l, r := int64(2), num
		for l < r {
			mid := l + (r-l)/2
			tmp := check(mid, i)
			if tmp == 0 {
				return strconv.FormatInt(mid, 10)
			} else if tmp < 0 {
				l = mid + 1
			} else {
				r = mid
			}
		}
	}

	return strconv.FormatInt(num-1, 10)
}
